flowauth/frontend/src/util/util.js

```diff
@@ -26,80 +27,144 @@ function zip() {
+  */
+export function scopesGraph(scopes_obj) {
+  var scopes_array
+  if (!Array.isArray(scopes_obj)){
+    scopes_array = Object.keys(scopes_obj)
+  }
+  else{
+    scopes_array = scopes_obj
+  }
+  const triplet_scopes = scopes_array.filter(row => row.includes(":"))  // It's this or a regex
+  const split_scopes = triplet_scopes.map(row => row.split(":"))
+  const scopes_tree = {}
+  for (const row of split_scopes){
+    var [ admin_level, tl_query, inner_query ] = row
+
+    if (!(Object.keys(scopes_tree).includes(admin_level))){
+      scopes_tree[admin_level] = {}
+    }
+    if (!(Object.keys(scopes_tree[admin_level]).includes(tl_query))){
+      scopes_tree[admin_level][tl_query] = {}
+    }
+    scopes_tree[admin_level][tl_query][inner_query] = true
+  }
+  const single_scopes=scopes_array.filter(row => !row.includes(":"))
+    .map(scope => ({[scope]:true}))
+    .reduce((p_x, x) => ({...p_x, ...x}))
+  const all_scopes = ({...scopes_tree, ...single_scopes})
+  console.log(all_scopes)
+  return all_scopes
 }
 
-export function jsonify(tree, labels, enabled, enabledKeys) {
-  const parent = tree.parent;
-
-  const list = Object.keys(tree)
-    .filter((k) => k !== "parent" && k !== "full_path")
-    .map((k) => {
-      const ll = labels.concat([k]);
-      const v = tree[k];
-      const val = parent ? [parent, ll.join(".")].join("&") : ll.join(".");
-      if (
-        Object.keys(v).filter((k) => k !== "parent" && k !== "full_path")
-          .length === 0
-      ) {
-        const value = [parent, v.full_path].join("&");
-        return {
-          label: k,
-          value: value,
-          parents: parent,
-          enabled: enabled.includes(value),
-        };
-      } else {
-        const children = jsonify(
-          v,
-          v.parent === parent ? ll : [],
-          enabled,
-          enabledKeys
-        );
-        const allEnabled = children.every((child) => child.enabled);
-        if (!allEnabled) {
-          children
-            .filter((child) => child.enabled)
-            .forEach((child) => enabledKeys.push(child.value));
-        }
-        return {
-          label: k,
-          value: val,
-          children: children,
-          parents: parent,
-          enabled: allEnabled,
-        };
+
+function jsonify_inner(tree, label, value, enabled){
+  const things = Object.keys(tree).map((branch, index) => {
+    const this_label = label ==="" ? branch: [label, branch].join(".")
+    const this_value = this_label
+    // const this_value = value === "" ? index.toString(): [value, index.toString()].join("-")
+    if (typeof(tree[branch]) === "boolean"){
+      if (tree[branch]){
+        enabled.push(this_value)
       }
-    });
-  if (parent === "") {
-    list
-      .filter((obj) => obj.enabled)
-      .forEach((obj) => enabledKeys.push(obj.value));
-  }
-  return list;
+      return {
+        label: this_label,
+        value: this_value,
+        enabled: this_value in enabled
```

Comment:
Why does this need to set an 'enabled' key? Seems to me that the initial checkbox states are set by the `value` input to RightsCascade, so the jsonified graph just needs to capture the nesting structure.

Reply:
A little bit of pre-engineering - I'm thinking that at some point, we may want to be able to turn scopes on or off from the server. Capturing `enabled` lets us filter here if it's turned on or off (iirc, everything is `enabled` by default atm, but this could change)